Adding ORANSlice framework (#34)
open-5g/openrangym: _posts/tutorials/2025-02-04-oranslice-tutorial.md

---
layout: post
title: "Deploying ORANSlice: A Comprehensive 5G Network Slicing Tutorial"
date: 2025-01-24
category: tutorials
author:
short-description: "How to deploy and run the ORANSlice framework for 5G network slicing integrating OAI core, RAN, and Near-RT RIC"
---

{% assign publications = site.data.publications %}

{% assign el = publications.publications | where: "name", "cheng2024oranslice" %}
{% assign element = el[0] %}
{% capture pub_cheng2024oranslice %}
{% include pub-template.html %}
{% endcapture %}

ORANSlice is an end-to-end open-source 5G network slicing framework for O-RAN. 

> It (i) extends the 5G protocol stacks of OAI to support RAN slicing; (ii) implements an E2SM-CCC-based SM and xApp for RAN slicing control; (iii) conducts extensive testing and validation on [Arena](https://openrangym.com/experimental-platforms/arena) and [X5G](https://x5g.org/) testbeds to ensure the robustness and effectiveness of the implementation.

Reviewer: Please open the X5G external link in a new tab

> It (i) extends the 5G protocol stacks of OAI to support RAN slicing; (ii) implements an E2SM-CCC-based SM and xApp for RAN slicing control; (iii) conducts extensive testing and validation on <a href="https://openrangym.com/experimental-platforms/arena" target="_blank">Arena</a> and <a href="https://x5g.org/" target="_blank">X5G</a> testbeds to ensure the robustness and effectiveness of the implementation.

Reviewer: ```suggestion
It (i) extends the 5G protocol stacks of OAI to support RAN slicing; (ii) implements an E2SM-CCC-based SM and xApp for RAN slicing control; (iii) conducts extensive testing and validation on [Arena](/experimental-platforms/arena) and <a href="https://x5g.org/" target="_blank">X5G</a> testbeds to ensure the robustness and effectiveness of the implementation.
```

Reviewer: Please use relative link for internal pages.

> More details can be found in the ORANSlice [GitHub page](https://github.com/wineslab/ORANSlice) and in the following paper:

Reviewer: Please open external link in a new tab

> {{ pub_cheng2024oranslice | strip_newlines }}
> {: .text-justify}

# Deployment Tutorial

There are three parts in an end-to-end 5G network slicing framework: core network, radio access network, and near-real-time RIC. 
The ORANSlice repo consists of the code of OAI RAN and OAI CN. The repo of near-real-time RIC is in other repos.
In this tutorial, we will show how to deploy the three parts in Ubuntu 22.04.

```
git clone https://github.com/wineslab/ORANSlice.git
```

## 5G Core Network Slicing

We provide OAI and Open5GS core network slicing deployment in docker.

### OAI Core 

> Go to folder `oai_cn` and follow the readme to deploy OAI core legacy version (v1.5.1) or latest version (develop branch).

Reviewer 2: ```suggestion
Deploy OAI core legacy version (`v1.5.1`) or latest version (`develop` branch) from the `oai_cn` directory. A more indepth guide is present in the the `README.md`.
```

### Open5GS

> The code of depolying Open5GS core network in docker is given in [Github Repo](https://github.com/wineslab/docker_open5gs/tree/open5gs_slicing). 

Reviewer: Please open external link in a new tab

> The code of depolying Open5GS core network in docker is given in [Github Repo](https://github.com/wineslab/docker_open5gs/tree/open5gs_slicing). 

Reviewer 2: ```suggestion
The code for deploying the Open5GS core network in Docker is available in the [Github repository](https://github.com/wineslab/docker_open5gs/tree/open5gs_slicing). 
```

## 5G RAN Slicing

The implementation of RAN Slicing is based on the OAI 5G protocal stack. 

### OAI monolithic gNB

#### Install protobuf-c for E2Agent
```
sudo apt-get update
sudo apt install protobuf-compiler
sudo apt install libprotoc-dev
git clone https://github.com/protobuf-c/protobuf-c && cd protobuf-c/ && ./autogen.sh && ./configure && make && sudo make install && sudo ldconfig
```

Go to folder `oai_ran` and follow the steps to build:
#### Build OAI gNB
```
cd cmake_targets
./build_oai -I
./build_oai -w USRP --ninja --gNB
```

The customized RAN slicing and E2 Agent code are integrated into OAI 2024.w28 release.

### Disaggregated gNB from NVIDIA Aerial and OAI 

> Please refer to the tutorial in [ARC-OTA](https://docs.nvidia.com/aerial/aerial-ran-colab-ota/current/) and [Aerial CUDA-Accelerated RAN](https://docs.nvidia.com/aerial/cuda-accelerated-ran/index.html).

Reviewer: Please open external link in a new tab

### OAI nrUE with multiple-PDU support
Our customized OAI nrUE with multiple-PDU support is given in 
> https://gitlab.eurecom.fr/oai/openairinterface5g/-/tree/NR_UE_multi_pdu

> Follow the [official guide](https://gitlab.eurecom.fr/oai/openairinterface5g/-/blob/NR_UE_multi_pdu/doc/NR_SA_Tutorial_OAI_nrUE.md) to build and run OAI nrUE.

Reviewer: Please open external link in a new tab

> Follow the [guide](https://gitlab.eurecom.fr/oai/openairinterface5g/-/blob/NR_UE_multi_pdu/doc/RUNMODEM.md#at-command-interface-in-nr-ue) to configure multiple PDU session at nrUE.

Reviewer: Please open external link in a new tab

> Follow the [guide](https://gitlab.eurecom.fr/oai/openairinterface5g/-/blob/NR_UE_multi_pdu/doc/RUNMODEM.md#at-command-interface-in-nr-ue) to configure multiple PDU session at nrUE.

Reviewer 2: ```suggestion
Follow the [guide](https://gitlab.eurecom.fr/oai/openairinterface5g/-/blob/NR_UE_multi_pdu/doc/RUNMODEM.md#at-command-interface-in-nr-ue) to configure multiple PDU sessions at nrUE.
```

**If you do not need multiple-pdu support at nrUE, you can build and run nrUE at `oai_ran`**

### COTS UE
You can use COTS 5G UE modules from Sierra Wireless or Quectel to establish network slicing connection.

## Near-Real-Time RIC

### OSC Near-RT RIC 

> The O-RAN Software Community (OSC) Near-RT RIC Release E can be installed following this [guide](https://docs.o-ran-sc.org/projects/o-ran-sc-ric-plt-ric-dep/en/latest/installation-guides.html). Note that for this tutorial, the OSC Near-RT RIC runs on an OpenShift cluster, whereas a plain Kubernetes deployment might require specific networking configurations.

Reviewer: Please open external link in a new tab

### E2Sim 

> The e2sim is a component that can run on the same server as the gNB and has the duty to encapsulate/decapsulate the custom Service Model (SM) buffers to be sent to or received from the gNB. It communicates with the gNB via UDP sockets. The e2sim codebase can be found at this GitHub [Repo](https://github.com/wineslab/o-ran-e2sim/tree/x5g-e2sim).

Reviewer: Please open external link in a new tab

### xApp

Our xAPP consists of two parts: a C/C++ based xAPP Base Station (BS) Connector and a python-based xAPP logic unit. The code can be found at the 

> [Github Repo](https://github.com/wineslab/xapp-oai/tree/oranslice).

Reviewer: Please open external link in a new tab
